### DISCUSSION TITLE
How to ceate production build without database connection?

### DISCUSSION THREAD

**Discussion by tjhoo** [vercel/next.js#67798] [Help]
### Summary

Hi,
I have a `./db/index.js` which initialize the database pool and export the client,
```
export const connection = await mysql.createConnection({ ... });
export const db = drizzle(connection, { ... });
```
then the database client is being imported in a server action `./app/actions.js` like this,
```
"use server";

import db from "@/db";

export const getStudents = () => { ... };
```
and the server action is imported in a page component
```
import { getStudents } from "./actions";

export default function Home() {
  return undefined
}
```
The  `npm run build` complaints the following error when there is no database server running,
```
warn - No utility classes were detected in your source files. If this is unexpected, double-check the `content` option in your Tailwind CSS configuration.
warn - https://tailwindcss.com/docs/content-configuration
 ✓ Compiled successfully
 ✓ Linting and checking validity of types
   Collecting page data  ..Error: connect ECONNREFUSED 127.0.0.1:3306
   ...
```
How can I create the build without the database connection?

**Comment by icyJoseph** [comment]
You need to use a factory pattern.

Change your db to something like:

```ts
// pseudocode...

let client;

const getConnection = () => {
  if (client) return client;

  const connection = await mysql.createConnection({ ... });
  const db = drizzle(connection, { ... });

  client = db;
  
  return client;

}
```

The intention is to defer the creation of the connection and the db instance, until you actually do need it, and once that's done, keep a reference to that instance, and serve the same instance on further calls to `getConnection`. So that your build step doesn't need to allocate an instance that it won't use, and that will fail because the db is not open to receive a connection.

**Comment by klclee** [comment]
@icyJoseph @tjhoo 

so am having similar problem using Nextjs latest (15). For example my Page.tsx is making a db call on render which is using the factory. next build actually invoke this during build time. 

same for next-auth setup when using a db adapter. It invokes the factory at build time. 

The issue for me is not having access to the env variable at build time. Thats next build, if you use trubo repo and set the "env" config in the build phase that works.

I am not sure just solving the ability to provide env at build is a good solution, somehow the default build should not invoke to to db connection at all. 

the way i got around this is using `next build --experimental-build-mode compile` think that should be the default tbh.
